docs: migrating lb3 model property mixins, and custom model method/remote method mixins to lb4

emonddr opened the pull request:
Related to : https://github.com/strongloop/loopback-next/issues/3951

<!--
Please provide a high-level description of the changes made by your pull request.

Include references to all related GitHub issues and other pull requests, for example:

Fixes #123
Implements #254
See also #23
-->

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

raymondfeng commented:
@emonddr Maybe we should create an example under `examples/migration-lb3-mixin` to host your code?

derdeka commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+}
+
+@model()
+export class Note extends addCategoryPropertyMixin(TempNote) {
```
```suggestion
export class Note extends addCategoryPropertyMixin(Entity) {
```
Can you shortly explain why there is a need to rename and use `TempNote` instead of using a fully implemented `Note` extending `Entity` which was exhanced by the mixin?

bajtos replied:
+1, I wanted to suggest the same. Take a look at how we are using mixins for Application classes:

https://github.com/strongloop/loopback-next/blob/28ebb8b7276577494d398423104d124b80241d43/examples/todo/src/application.ts#L15-L17

Ideally, I'd like use to describe the same approach for model mixins too. Here is an illustration of what I mean:

```ts
@model()
export class Note extends addCategoryPropertyMixin(Entity) {
  @property({
    type: 'number',
    id: true,
    generated: true,
  })
  id?: number;

  // etc.
}
```

Unless there are reasons why that does not work and the mixin must be applied on the class containing other model properties?

emonddr replied:
The mixin requires something with a constructor. And `Entity` is an abstract class.

```
export abstract class Entity extends Model implements Persistable 
```
https://github.com/strongloop/loopback-next/blob/master/packages/repository/src/model.ts#L276

emonddr replied:
hmm... the only way I was able to get it working was by doing this:
```
import {Entity, model, property} from '@loopback/repository';
import {addCategoryPropertyMixin} from '../mixins/categoryPropertyMixin';

class BaseClass extends Entity {};

@model()
export class Note extends addCategoryPropertyMixin(BaseClass) {
  constructor(data?: Partial<Note>) {
    super(data);
  }

  @property({
    type: 'number',
    id: true,
    generated: true,
  })
  id?: number;

  @property({
    type: 'string',
    required: true,
  })
  title: string;

  @property({
    type: 'string',
  })
  content?: string;

}


export interface NoteRelations {
  // describe navigational properties here
}

export type NoteWithRelations = Note & NoteRelations;

```

emonddr replied:
@raymondfeng suggested to me that we can make Entity a normal class (non-abstract)

derdeka replied:
> @raymondfeng suggested to me that we can make Entity a normal class (non-abstract). Then I don't need to create BaseClass at all. Just pass in Entity.

+1, but I guess this will introduce some side effects.

I just remembered that i stumbled on this a year ago when using model mixins in my applications. I also introduced a `BaseModel` but with common attributes for all my models.

There is a similar issue when applying mixins on `ValueObject`.

emonddr replied:
@derdeka , so you are ok with 

```
class BaseModel extends Entity {};

...

export class Note extends addCategoryPropertyMixin(BaseModel)

```

approach

and no longer ok with 

```
export class Note extends addCategoryPropertyMixin(Entity)
```
if it means making `Entity` non-abstract?

What kind of side-effects on Entity did this cause? I will run regression tests surrounding it to make sure there are no regressions (or side effects), but if you could elaborate, that would be helpful. thx.

derdeka replied:
Both approaches are fine to me.

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+
+Before we use the mixin class factory function:
+
+- rename this class to `TempNote`
```
The name `TempNote` suggests to me that this is just a temporary thing that will be removed later.  That's not true, is it?

Let's use a more descriptive class name than `TempNote` please, for example `BaseNote`.

emonddr replied:
> The name TempNote suggests to me that this is just a temporary thing that will be removed  later. That's not true, is it? Let's use a more descriptive class name than TempNote please, for example BaseNote. 

No longer using TempNote

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+
+import {property} from '@loopback/repository';
+
+export function addCategoryPropertyMixin<T extends Constructor<any>>(
```
We usually use PascalCase naming for mixins applied on class constructors, see e.g. https://github.com/strongloop/loopback-next/blob/28ebb8b7276577494d398423104d124b80241d43/examples/todo/src/application.ts#L15-L17

```suggestion
export function AddCategoryPropertyMixin<T extends Constructor<any>>(
```

emonddr replied:
@bajtos , I did look at how Mixins were used in lb4, and at the moment, they only center around the main application. :)

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+- updating the model's json file to include the mixin's name (and options object
+  or boolean)
+
+### Defining The Model Property Mixin findByTitle.js
```
I think this should say "Model Method Mixin"?

```suggestion
### Defining The Model Method Mixin findByTitle.js
```

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+        title: title,
+      },
+    };
+    Model.find(titleFilter, function(err, foundItems) {
```
Since your callback is just forwarding the results to `cb`, there is no need to write it and you can pass `cb` directly to `find`.

```suggestion
    Model.find(titleFilter, cb);
```

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+- the locations of all **models**
+- the location of all **mixins**
+- the entry of the model that receives the mixin content (for this example Note)
+
```
I think these instructions are the same as for Model Property mixins, can we remove them to avoid duplication and point readers to the content for Model Property mixins?

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+
+- placing the mixin logic in a file in a mixins directory
+- updating the server/model-config.json file with the mixin directory location
+- updating the model's json file to include the mixin's name and a boolean
```
I love ❤️ these short list of steps that you are showing in different places of the document, they make it easier to grasp the overal structure of this guide 👍

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+import {Model} from '@loopback/repository';
+import {FindByTitleInterface} from './findByTitleInterface';
+
+export function FindByTitleRepositoryMixin<
```
Kudos for using PascalCase for the mixin name :+1:

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+
+export function FindByTitleRepositoryMixin<
+  M extends Model,
+  T extends Constructor<any>
```
Is it possible to clarify these generic arguments to make it more clear that T is expected to be a class implementing `CrudRepository<M>`? 

Ideally, this should be expressed in a way that allows the TypeScript compiler to verify the contract at compile time. If that's not possible or too difficult, then at least add a short comment to explain what is `T` for. 

I also think that `R` would be a better name (`R` as a shortcut for `Repository`).

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+import {DbDataSource} from '../datasources';
+import {inject, Constructor} from '@loopback/core';
+
+export class NoteRepository extends FindByTitleRepositoryMixin<
```
Kudos for using the pattern `class MyClass extends MyMixin(BaseClass)` 👍

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+export function FindByTitleControllerMixin<
+  M extends Model,
+  T extends Constructor<any>
+>(superClass: T, parms: FindByTitleControllerMixinParms) {
```
I believe we usually abbreviate `parameters` as `params` (i.e. `params` argument, `FindByTitleControllerMixinParams` interface).

Having said that, we tend to use the name `options` instead of `params`.

```suggestion
>(superClass: T, options: FindByTitleControllerMixinOptions) {
```

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+    @get(parms.basePath + '/findByTitle/{title}', {
+      responses: {
+        '200': {
+          description: `Array of ${parms.modelClassName} model instances`,
```
I believe `modelClassName` argument is redundant, you can get the model name via `parms.modelClass.modelName`.

https://github.com/strongloop/loopback-next/blob/abc4c07863060921abf965af3f29718c0caab0b8/packages/repository/src/model.ts#L203-L206

emonddr replied:
Hmmm. It doesn't work

![image](https://user-images.githubusercontent.com/6864736/75309925-739ba780-5820-11ea-9169-33dfb53acac0.png)

emonddr replied:
@bajtos

bajtos replied:
I wish there was a runnable code I could use to try this out myself and tinker with the types.

I think the problem is caused by the fact your `modelClass` argument is specified as `Constructor<object>`, which means any class can be provided as the value and therefore we cannot guarantee that `modelName` static property is available.

IMO, this is a bug, `FindByTitleControllerMixin` should allow only Model constructors. Here is the type definition that should (hopefully) work:

```ts
export interface FindByTitleControllerMixinOptions {
  basePath: string;
  modelClass: typeof Model;
}
```

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+    async findByTitle(@param.path.string('title') title: string): Promise<[M]> {
+      let foundItems = await this.repository.findByTitle(title);
+
+      return foundItems;
```
Nitpick: the method body can be simplified as follows:

```ts
class MixedController  /*...*/ {
  
  @get(parms.basePath + '/findByTitle/{title}', {
    // ...
  })
  findByTitle(@param.path.string('title') title: string): Promise<[M]> {
    return this.repository.findByTitle(title);
  }
}
```

This is not only less lines of code, but also slightly more performant, because it creates one less Promise object at runtime.

See https://bajtos.net/2018-AsyncAwait/#/9/6 and [promise rewrapping](https://exploringjs.com/es2016-es2017/ch_async-functions.html#_returned-promises-are-not-wrapped)

bajtos commented on docs/site/migration/models/mixins.md:
````diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+{% include code-caption.html content="src/controllers/note.controller.ts" %}
+
+```ts
+class TempNoteController {
````
Same comments as above:

1. Don't use `Temp` in the name, find a better prefix/postfix please. E.g. `NoteControllerBase` or `BaseNoteController`.
2. Even better, don't create any temporary classes, apply the mixin on the base class of the controller. In this particular case, I think you should use `Object` as the base.

```ts
class NoteController extends extends FindByTitleControllerMixin<
   Note,
   Constructor<TempNoteController>
 >(Object, mixinParms) {}
```

jannyHou commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+          title: title,
+        },
+      };
+      const foundItems = await this.find(titleFilter);
```
a question...if the mixin doesn't have a function `find`, how could it call `find` here?

emonddr replied:
@jannyHou `find` is a function available on all models in lb3.

jannyHou replied:
oh I didn't mean lb3, I mean, typescript doesn't complain sth like `find` not found?

emonddr replied:
no transpilation errors.

emonddr replied:
@raymondfeng , any idea why there is no complaint of `this.find()`?

I know this code works:

```
export function FindByTitleRepositoryMixin<
  M extends Model,
  R extends Constructor<any>
>(superClass: R) {
  return class extends superClass implements FindByTitleInterface<M> {
    async findByTitle(title: string): Promise<M[]> {
      const titleFilter = {
        where: {
          title: title,
        },
      };
      return await this.find(titleFilter);
    }
  };
}
```
But just curious why ts doesn't complain.

bajtos replied:
Because `R` is defined as `Constructor<any>`, your super class is effectively extending from `any`. If you hover your mouse over `this.find` in VS Code, you will see that it has `any` type.

<img width="621" alt="Screen Shot 2020-02-27 at 14 34 46" src="https://user-images.githubusercontent.com/1140553/75450007-52e16800-596e-11ea-8a6f-c6b73a8d8121.png">

Here is a more correct definition of `FindByTitleRepositoryMixin`:

```ts
export function FindByTitleRepositoryMixin<
  M extends Model,
  R extends Constructor<CrudRepository<M>>
>(superClass: R) {
  return class extends superClass implements FindByTitleInterface<M> {
    // ...
  };
}
```

I am choosing `CrudRepository` because that's the interface where `find` is defined. By using that interface instead of e.g. `DefaultCrudRepository`, we make it possible to use this mixin with other repository implementations too.

bajtos replied:
The problem with the code above is that TS now complains that `title` cannot be used in a filter, because a generic `Model` does not necessarily have a filter property. The solution is to limit `M` to `extends Model & {title: string}`.

```ts
export function FindByTitleRepositoryMixin<
  M extends Model & {title: string},
  R extends Constructor<CrudRepository<M>>
>(superClass: R) {
```

Unfortunately that's still not enough, because our `Condition<MT>` type cannot correctly handle the case where `MT` is a generic argument extending `Model & {title: string}`. I am proposing to use the following implementation:

```ts
    async findByTitle(title: string): Promise<M[]> {
      const where = {title} as Where<M>;
      const titleFilter = {where};
      return this.find(titleFilter);
    }
```

/cc @raymondfeng would you like to take a look on why `{title}` cannot be assigned to `Condition<M>` where `M` is a generic argument defined as `M extends Model & {title: string}`?

Example code to reproduce:

```ts
function test<M extends Model & {title: string}>(): Condition<M> {
  return {title: 'foo'};
}
```

Compiler error:

```
Type '{ title: "foo"; }' is not assignable to type 'Condition<M>'.ts(2322)
```

jannyHou commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+
+Import the `FindByTitleRepositoryMixin` mixin class factory function, and adjust
+the declaration of the `NoteRepository` class to extend the class returned from
+the mixin function that takes in the `DefaultCrudRepository` superclass an
```
nitpick: an --> as?

emonddr commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+    }
+  },
+  "mixins": {
+    "Category": true
```
Not sure why, but when entering mixin information, things are in uppercase on first letter.
As in 👍 
https://github.com/strongloop/loopback-example-mixins/blob/master/common/models/dog.json#L26

```
  "mixins": {
    "TimeStamp": true,
    "Tag": true,
    "Squirrel": true
  }
```
all the mixin files are in lower case, but when specifying the info in  dog.json, they must be with first letter uppercase. 

@bajtos , @raymondfeng  , any ideas why this is?

jannyHou replied:
I see then I guess the loopback-boot has some name mapping happen behind the scene.

emonddr commented on docs/site/migration/models/mixins.md:
````diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+
+{% include code-caption.html content="server/model-config.json" %}
+
+```
````
I can't because when I do that, it makes my comment

```
// ... other entries
```
all red and weird

jannyHou replied:
Oh you make a good point, JSON doesn't support comment. 👍 never mind, I am ok with either, w/ or w/o the type
https://stackoverflow.com/questions/244777/can-comments-be-used-in-json

emonddr commented on docs/site/migration/models/mixins.md:
````diff
@@ -6,8 +6,805 @@ sidebar: lb4_sidebar
+```
+{
+  "_meta": {
+    "sources": [
````
This is a typical setup for mode-config.json as explained in 👍 
https://loopback.io/doc/en/lb3/Defining-mixins.html#reference-mixins-in-model-configjs

I will make a reference to this for an explanation.

emonddr commented:
I still need to refactor the note, and controller classes as suggested above, but made most suggested changes in 2nd commit.

emonddr commented:
Running test suite to ensure no regressions by making Entity class non-abstract....

raymondfeng commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,699 @@ sidebar: lb4_sidebar
+export class NoteRepository extends FindByTitleRepositoryMixin<
+  Note,
+  Constructor<
+    DefaultCrudRepository<Note, typeof Note.prototype.id, NoteRelations>
```
Can it be as simple as `Constructor<Note>` or `typeof Note`?

emonddr replied:
> Can it be as simple as `Constructor<Note>` or `typeof Note`?

Can you give an example of what you mean? thx. :)

raymondfeng replied:
To replace `DefaultCrudRepository<Note, typeof Note.prototype.id, NoteRelations>`

emonddr replied:
@raymondfeng , I cannot get this to work in the suggested manner.

raymondfeng commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,699 @@ sidebar: lb4_sidebar
+
+export interface FindByTitleControllerMixinOptions {
+  basePath: string;
+  modelClass: Constructor<{}>;
```
`Constructor<object>`?

raymondfeng commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,699 @@ sidebar: lb4_sidebar
+      super(...args);
+    }
+
+    @get(options.basePath + '/findByTitle/{title}', {
```
It's better to use:
```ts
@get(`${options.basePath}/findByTitle/{title}`, {
```

raymondfeng commented on packages/repository/src/model.ts:
```diff
@@ -273,7 +273,7 @@ export abstract class ValueObject extends Model implements Persistable {}
  * Base class for entities which have unique ids
  */
-export abstract class Entity extends Model implements Persistable {
+export class Entity extends Model implements Persistable {
```
Can you have a separate PR for this? Is it still required with `BaseModel`?

emonddr replied:
> Can you have a separate PR for this? Is it still required with `BaseModel`?

I just snuck the commit in there so I can run the CI. I ran the tests locally but they keep dying on http proxy and example.com .  This change is not currently part of my proposed solution. I use BaseModel extends Entity {} right now.

raymondfeng replied:
If so, please revert the change.

emonddr replied:
Reverted the change. No longer exploring the idea of making `Entity` non-abstract.

bajtos commented on docs/site/migration/models/mixins.md:
````diff
@@ -116,13 +121,30 @@ the mixin.
+
+```ts
+import {Entity} from '@loopback/repository';
+export class BaseModel extends Entity {}
````
I find it confusing that the class is called `BaseModel` when it's extending from `Entity`. What if the user wants to implement a mixin that can be applied to `Model` subclasses too, not only to `Entity` descendants? 

I am proposing to use the name `BaseEntity` instead.

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -566,30 +505,30 @@ import {Model} from '@loopback/repository';
+export interface FindByTitleControllerMixinOptions {
   basePath: string;
-  modelClass: Constructor<{}>;
+  modelClass: Constructor<object>;
```
As mentioned in https://github.com/strongloop/loopback-next/pull/4723#discussion_r383925889, I think this should be:

```suggestion
  modelClass: typeof Model;
```

bajtos commented on docs/site/migration/models/mixins.md:
```diff
@@ -566,30 +505,30 @@ import {Model} from '@loopback/repository';
   basePath: string;
-  modelClass: Constructor<{}>;
+  modelClass: Constructor<object>;
   modelClassName: string;
```
Let's find a way how to avoid the requirement to specify both model class & model name, see the discussion in https://github.com/strongloop/loopback-next/pull/4723#discussion_r383925889

bajtos commented:
> I think we should introduce an automated test to put the proposed mixin code snippets together and verify that the code can be compiled and works as intended at runtime. See the #4723 for inspiration. That way when we need to tweak the migration instructions later in the future, we have a place where to try the proposed modifications and verify that they work as intended.

Just to make it clear, I am asking for tests verifying the mechanics of implementing a LB4 mixin class and applying it to a model, a repository and a controller. These tests SHOULD NOT reproduce an entire end-to-end LB4 application with its different source code files and so on. The migration guide tells users to put different pieces of code to different source-code files. In a test, all pieces can be together in the test file (or a fixture file, if that's more appropriate).

See [Define your testing strategy](https://loopback.io/doc/en/lb4/Defining-your-testing-strategy.html#define-your-testing-strategy) and [How to build a great test suite](https://loopback.io/doc/en/lb4/Defining-your-testing-strategy.html#how-to-build-a-great-test-suite) in our docs.

emonddr commented:
@bajtos 

Using 
```
export function FindByTitleRepositoryMixin<
  M extends Model & {title: string},
  R extends Constructor<CrudRepository<M>>
>(superClass: R) {
```

and

```
   async findByTitle(title: string): Promise<M[]> {
      const where = {title} as Where<M>;
      const titleFilter = {where};
      return this.find(titleFilter);
    }
```

works

thx

emonddr commented:
Just had a short meeting with @bajtos , and we agreed that I should create some model and repo mixin test cases in the package/repository  and a controller mixin in package/rest .

An examples/mixins can be a separate PR and task.

emonddr commented:
> > I think we should introduce an automated test to put the proposed mixin code snippets together and verify that the code can be compiled and works as intended at runtime. See the #4723 for inspiration. That way when we need to tweak the migration instructions later in the future, we have a place where to try the proposed modifications and verify that they work as intended.
> 
> Just to make it clear, I am asking for tests verifying the mechanics of implementing a LB4 mixin class and applying it to a model, a repository and a controller. These tests SHOULD NOT reproduce an entire end-to-end LB4 application with its different source code files and so on. The migration guide tells users to put different pieces of code to different source-code files. In a test, all pieces can be together in the test file (or a fixture file, if that's more appropriate).
> 
> See [Define your testing strategy](https://loopback.io/doc/en/lb4/Defining-your-testing-strategy.html#define-your-testing-strategy) and [How to build a great test suite](https://loopback.io/doc/en/lb4/Defining-your-testing-strategy.html#how-to-build-a-great-test-suite) in our docs.

I will work on these tests today.

emonddr commented:
I also created this followup doc task suggested by @bajtos : https://github.com/strongloop/loopback-next/issues/4759

emonddr commented:
@bajtos , @raymondfeng , I have added tests to prove that mixins in docs work.

dhmlau commented:
A suggestion to consider: Since it's a migration guide, instead of we have:
```
LoopBack 3 approach to .....
LoopBack 4 approach to .....
```

would it be better to have:
```
Creating a Property Mixin
   LoopBack 3 approach
   LoopBack 4 approach
Creating A Custom Model Method And Remote Model Method Mixin
   LoopBack 3 approach
   LoopBack 4 approach
```

bajtos commented on packages/repository/src/__tests__/fixtures/mixins/categoryPropertyMixin.ts:
```diff
@@ -0,0 +1,26 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
```
We use kebab-case in file names. Please rename this file to `category-property-mixin.ts`.

bajtos commented on packages/repository/src/__tests__/fixtures/mixins/categoryPropertyMixin.ts:
```diff
@@ -0,0 +1,26 @@
+// License text available at https://opensource.org/licenses/MIT
+
+import {Constructor} from '@loopback/context';
+import {property} from '../../../index';
```
Please either remove `index` (import from `'../../..`), or change the import to reference the specific file implementing the decorator (I think that would be `'../../../src/decorators/model.decorator`).

bajtos commented on packages/repository/src/__tests__/fixtures/mixins/categoryPropertyMixin.ts:
```diff
@@ -0,0 +1,26 @@
+import {property} from '../../../index';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function AddCategoryPropertyMixin<T extends Constructor<any>>(
```
Can we use `Constructor<Entity>` to avoid no-explicit-any problem? I think such type is more correct than `Constructor<any>`.

bajtos commented on packages/repository/src/__tests__/fixtures/mixins/findByTitleInterface.ts:
```diff
@@ -0,0 +1,10 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
```
Please rename this file to `find-by-title-interface.ts`.

Personally, I find it too verbose to have a dedicated file for this interface, have you considering moving the interface to the same file where `FindByTitleRepositoryMixin` is defined?

bajtos commented on packages/repository/src/__tests__/fixtures/mixins/findByTitleInterface.ts:
```diff
@@ -0,0 +1,10 @@
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Model} from '../../../index';
```
Please remove `index` from the import path.

bajtos commented on packages/repository/src/__tests__/fixtures/mixins/findByTitleRepoMixin.ts:
```diff
@@ -0,0 +1,25 @@
+// License text available at https://opensource.org/licenses/MIT
+
+import {Constructor} from '@loopback/context';
+import {CrudRepository, Model, Where} from '../../../index';
```
Please remove `index` from the import path.

bajtos commented on packages/repository/src/__tests__/fixtures/mixins/findByTitleRepoMixin.ts:
```diff
@@ -0,0 +1,25 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
```
Please rename the file to `find-by-title-repo-mixin.ts`.

bajtos commented on packages/repository/src/__tests__/unit/mixins/add-method-to-repository.mixin.unit.ts:
```diff
@@ -0,0 +1,48 @@
+  it(`mixin method 'findByTitle' exists`, () => {
+    expect('findByTitle' in repo).to.be.True();
+    expect(typeof repo.findByTitle === 'function').to.be.True();
+  });
```
Can you please add a code calling `findByTitle`, to verify that TypeScript is seeing a correct signature? Something along the lines of:

```ts
const note: Note = await repo.findByTitle('a-title');
```

bajtos commented on packages/repository/src/__tests__/unit/mixins/model.mixin.unit.ts:
```diff
@@ -0,0 +1,48 @@
+
+describe('add property to model via mixin', () => {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  let meta: MetadataMap<any>;
```
Can we replace `any` with `unknown`?

bajtos commented on packages/rest/src/__tests__/fixtures/controllers/note.controller.ts:
```diff
@@ -0,0 +1,200 @@
+  })
+  async deleteById(@param.path.number('id') id: number): Promise<void> {
+    return Promise.resolve();
+  }
```
Do we really need all these CRUD controller methods for the test? Please keep only those methods that are relevant. I think that means either one method (to verify that both mixed-in and user-provided methods are recognized) or no methods at all (because we want to verify that mixed-in methods are added, we don't really care about user-provided methods).

bajtos commented on packages/rest/src/__tests__/fixtures/mixins/findByTitleControllerMixin.ts:
```diff
@@ -0,0 +1,50 @@
+export function FindByTitleControllerMixin<
+  M extends Model,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  T extends Constructor<any>
```
Can we use `Constructor` or `Constructor<Object>` to avoid `any` problem?

bajtos commented on packages/rest/src/__tests__/fixtures/mixins/findByTitleControllerMixin.ts:
```diff
@@ -0,0 +1,50 @@
+>(superClass: T, options: FindByTitleControllerMixinOptions) {
+  class MixedController extends superClass implements FindByTitleInterface<M> {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    constructor(...args: any[]) {
```
It makes me wonder if it's possible to use `unknown` instead of `any` here.

bajtos commented on packages/rest/src/__tests__/fixtures/mixins/findByTitleControllerMixin.ts:
```diff
@@ -0,0 +1,50 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
```
Please use kebab-case in the file name.

bajtos commented on packages/rest/src/__tests__/fixtures/mixins/findByTitleInterface.ts:
```diff
@@ -0,0 +1,10 @@
+
+export interface FindByTitleInterface<M extends Model> {
+  findByTitle(title: string): Promise<M[]>;
+}
```
Please rename this file to `find-by-title-interface.ts`.

Personally, I find it too verbose to have a dedicated file for this interface, have you considering moving the interface to the same file where `FindByTitleControllerMixin` is defined?

bajtos commented on packages/rest/src/__tests__/unit/controller-mixin/controller.mixin.unit.ts:
```diff
@@ -0,0 +1,38 @@
+    expect('findByTitle' in controller).to.be.True();
+    expect(typeof controller.findByTitle === 'function').to.be.True();
+  });
+});
```
It would be nice to have a test that's invoking the mixed-in endpoint via HTTP, to ensure it's correctly picked up by our REST layer.

bajtos commented on packages/repository/src/__tests__/fixtures/mixins/baseEntity.ts:
```diff
@@ -0,0 +1,11 @@
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Entity} from '../../../index';
```
Please either remove `index` (import from `'../../..`), or change the import to reference the specific file implementing the imported artifacts.

raymondfeng replied:
It should be `import {Entity} from '../../..`

raymondfeng commented on packages/repository/src/__tests__/fixtures/mixins/baseEntity.ts:
```diff
@@ -0,0 +1,11 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
```
The file name uses camel case, which is not consistent with our code base. I guess you rename it to `base-entity.ts`.

emonddr commented:
> Can we use Constructor<Entity> to avoid no-explicit-any problem? I think such type is more correct than Constructor<any>. 

I spoke with @bajtos  and we agreed I can use `Model` instead.

raymondfeng commented on packages/repository/src/__tests__/fixtures/mixins/findByTitleRepoMixin.ts:
```diff
@@ -0,0 +1,25 @@
+
+/*
+ * This function adds a new method 'findByTitle' to a repository class
+ * where 'M' is a model and 'R' is DefaultCrudRepository
```
Please use `@typeParam` tsdoc comments.

raymondfeng commented on packages/repository/src/__tests__/fixtures/mixins/findByTitleRepoMixin.ts:
```diff
@@ -0,0 +1,25 @@
+
+import {Constructor} from '@loopback/context';
+import {CrudRepository, Model, Where} from '../../../index';
+import {FindByTitleInterface} from './findByTitleInterface';
```
I think we can simply move `FindByTitleInterface` into this file. The interface name can be just `FindByTitle`.

raymondfeng commented on packages/repository/src/__tests__/fixtures/models/note.model.ts:
```diff
@@ -0,0 +1,39 @@
+
+import {model, property} from '../../../';
+import {BaseEntity} from '../../fixtures/mixins/baseEntity';
+import {AddCategoryPropertyMixin} from '../../fixtures/mixins/categoryPropertyMixin';
```
It should be:
```ts
import {BaseEntity} from '../mixins/baseEntity';
import {AddCategoryPropertyMixin} from '../mixins/categoryPropertyMixin';
```

raymondfeng commented on packages/repository/src/__tests__/fixtures/repositories/note.repository.ts:
```diff
@@ -0,0 +1,20 @@
+import {FindByTitleRepositoryMixin} from '../mixins/findByTitleRepoMixin';
+import {DefaultCrudRepository, juggler} from '../../..';
+import {Note, NoteRelations} from '../../fixtures/models/note.model';
+import {Constructor} from '@loopback/core';
```
It should be:

```ts
import {Constructor} from '@loopback/core';
import {DefaultCrudRepository, juggler} from '../../..';
import {FindByTitleRepositoryMixin} from '../mixins/findByTitleRepoMixin';
import {Note, NoteRelations} from '../models/note.model';
```

raymondfeng commented on packages/repository/src/__tests__/unit/mixins/add-method-to-repository.mixin.unit.ts:
```diff
@@ -0,0 +1,48 @@
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {NoteRepository} from '../../fixtures/repositories/note.repository';
```
Fix this!

raymondfeng commented on packages/repository/src/__tests__/fixtures/mixins/findByTitleRepoMixin.ts:
```diff
@@ -0,0 +1,25 @@
+export function FindByTitleRepositoryMixin<
+  M extends Model & {title: string},
+  R extends Constructor<CrudRepository<M>>
+>(superClass: R) {
```
It can be simplified as:
```ts
export function FindByTitleRepositoryMixin<M extends Model & {title: string}>(
  superClass: Constructor<CrudRepository<M>>,
) {
```

emonddr replied:
> It can be simplified as:
> 
> ```ts
> export function FindByTitleRepositoryMixin<M extends Model & {title: string}>(
>   superClass: Constructor<CrudRepository<M>>,
> ) {
> ```

It turns out your new proposal works at runtime, but the compiler is not aware of certain methods like repo.exists or repo.findById. 

This was discovered when @bajtos suggested tests for these methods too.

Then we re-worked the repo mixin to your original suggestion last week:

```
export function FindByTitleRepositoryMixin<
  E extends Entity & {title: string},
  ID,
  Relations extends object = {}
>(superClass: Constructor<EntityCrudRepository<E, ID, Relations>>) {
  return class extends superClass implements FindByTitle<E> {
    async findByTitle(title: string): Promise<E[]> {
      const where = {title} as Where<E>;
      const titleFilter = {where};
      return this.find(titleFilter);
    }
  };
}
```

emonddr commented:
@bajtos , @raymondfeng 

regarding

> constructor(...args: any[])

Changing this 

`constructor(...args: unknown[])`

gives the error

```
A mixin class must have a constructor with a single rest parameter of type 'any[]'.ts(2545)
```

emonddr commented:
Not sure what is wrong with this : https://github.com/strongloop/loopback-next/pull/4723#discussion_r385859394

raymondfeng commented:
> constructor(...args: any[])

We have to keep it.

raymondfeng commented on packages/rest/src/__tests__/unit/controller-mixin/controller.mixin.unit.ts:
```diff
@@ -0,0 +1,58 @@
+    return whenIMakeRequestTo(server)
+      .get('/notes/findByTitle/groceries')
+      .expect(200)
+      .expect([groceryNote.toJSON()]);
```
Nitpick: it can be merged as `.expect(200, [groceryNote.toJSON()]);`

raymondfeng commented on packages/rest/src/__tests__/unit/controller-mixin/controller.mixin.unit.ts:
```diff
@@ -0,0 +1,58 @@
+
+  it(`non mixin method 'create' exists`, () => {
+    expect('create' in controller).to.be.True();
+    expect(typeof controller.findByTitle === 'function').to.be.True();
```
It should be `expect(controller.findByTitle).to.be.a.function();`

raymondfeng commented on packages/rest/src/__tests__/unit/controller-mixin/controller.mixin.unit.ts:
```diff
@@ -0,0 +1,58 @@
+  });
+
+  it(`non mixin method 'create' exists`, () => {
+    expect('create' in controller).to.be.True();
```
It should be refined as `expect(controller).to.have.property('create');`

raymondfeng commented on packages/rest/src/__tests__/unit/controller-mixin/controller.mixin.unit.ts:
```diff
@@ -0,0 +1,58 @@
+
+  it(`mixin method 'findByTitle' exists`, async () => {
+    expect('findByTitle' in controller).to.be.True();
+    expect(typeof controller.findByTitle === 'function').to.be.True();
```
Ditto.

raymondfeng commented on packages/repository/src/__tests__/unit/mixins/add-method-to-repository.mixin.unit.ts:
```diff
@@ -0,0 +1,55 @@
+    ];
+    const methodsFound = crudMethods.filter(methodName => {
+      return methodName in repo;
+    });
```
```ts
const methodsFound = crudMethods.filter(methodName => typeof repo[methodName] === 'function');
```

emonddr replied:
> ```ts
> const methodsFound = crudMethods.filter(methodName => typeof repo[methodName] === 'function');
> ```

Cannot seem to use this...it gives an error:

```
Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'NoteRepository'.
  No index signature with a parameter of type 'string' was found on type 'NoteRepository'.ts(7053)
```

bajtos replied:
Re-posting from our chat:

Personally, I am not a big fan of checking exactly what methods an interface or a class has. Such tests are brittle, because when we add a new method to `DefaultCrudRepository`, then the test will start to fail. Surely the test is not interested in exact method names provided by `DefaultCrudRepository`? IIUC, you want to say "repo has all DefaultCrudRepository members plus `findByTitle`". So the correct solution is to list all members of DefaultCrudRepository, add `findByTitle` to that array, and use that as the expected list of members in `repo`.

---

Let's take a look at your code:

```ts
    const methodsFound = crudMethods.filter(methodName => {
      return methodName in repo && typeof repo[methodName] === 'function';
    });
```

It takes an array of strings (crudMethods) and for each string value, it tries to access a member of repo with that name.

The problem is that from TypeScript's point of view, those strings can be arbitrary - you can sneak in `explode` or make a typo and use `craete` instead of `create`.

When analyzing the expression `repo[methodName]`, TypeScript is not able to figure out what type does it resolves to. It's a valid statement in JavaScript and because TypeScript cannot figure out what's the type, it uses `any`.

The problem with `any` is that once you get `any`, all further compile checks are pretty much disabled. This is often difficult to spot, therefore we have configured the compiler to tell us about these situations and reject them with errors.

I am proposing to build the list of expected methods as follow (please add the typeof function check yourself):

```ts
    const crudMethods: (keyof DefaultCrudRepository<
      Note,
      typeof Note.prototype.id
    >)[] = Object.keys(DefaultCrudRepository);
```

With that change in place, the compiler will discover the problem described in my comment https://github.com/strongloop/loopback-next/pull/4723#pullrequestreview-367270278 below.

bajtos replied:
The code snippet above does not work, here is the corrected version:

```ts
    const crudProto = DefaultCrudRepository.prototype;
    const crudMembers = Object.keys(crudProto) as (keyof DefaultCrudRepository<
      Note,
      typeof Note.prototype.id
    >)[];
    const crudMethods = crudMembers.filter(
      key => typeof crudProto[key] === 'function',
    );
```

raymondfeng replied:
Good catch!

emonddr replied:
@bajtos thanks for the workaround. But I decided only to test `create`, `findById`, and `findByTitle` ; instead of all the crud functions.

emonddr replied:
@bajtos , actually after having a discussion with @raymondfeng , we decided to keep the using of `Model` instead of `Entity` for all the mixin functions, and use `CrudRepository` instead of `EntityCrudRepository` for the sole reason that the mixin method in the documentation example add a `findByTitle` method which only needs to use `repo.find`...it doesn't need `create` or `findById` nor `exists`. So our mocha tests will reflect this as well. I have reverted the repomixin changes, and will only test `create` and `findByTitle`.  :)

emonddr replied:
```
    const crudProto = DefaultCrudRepository.prototype;
    const crudMembers = Object.keys(crudProto) as (keyof DefaultCrudRepository<
      Note,
      typeof Note.prototype.id
    >)[];
    const crudMethods = crudMembers.filter(
      key => typeof crudProto[key] === 'function',
    );
```
This just produces empty arrays for me.

bajtos replied:
> we decided to keep the using of `Model` instead of `Entity` for all the mixin functions, and use `CrudRepository` instead of `EntityCrudRepository` for the sole reason that the mixin method in the documentation example add a `findByTitle` method which only needs to use `repo.find`...it doesn't need `create` or `findById` nor `exists`.

Sure, that makes a lot of sense. I my view, we are designing the mixing to require only those methods that we actually need in the mixin implementation. If `findByTitle` needs only the `find` method, then we should require `CrudRepository` only and don't ask for `EntityCrudRepository` methods that we don't need.

> will only test `create` and `findByTitle`. 

I am afraid I strongly disagree with the second part. When the user calls `FindByTitleRepositoryMixin` and provides a subclass of `DefaultCrudRepository` (which implements `EntityCrudRepository` interface), then the class created by the mixin should have all `EntityCrudRepository` methods. This will be always true at runtime, we need to find a way how to make the TypeScript compiler understand that at compile-time too.

A test calling `repo.findById` verifies both compile-time and run-time behavior, I think it's crucial to keep such test included.

Based on the newer comments and the the latest version presented in cbcc296, I think my concerns are actually addressed because the test for `findById` is present :+1:

emonddr commented:
> Re-posting from our chat:
> 
> > The compiler sees only `CrudRepository` members, it does not see members contributed by `EntityCrudRepository` on top of `CrudRepository` (e.g. `exists` and `findById`)
> 
> I think it is very important to capture this scenario in a test and make sure it works both at compile time and runtime too.
> 
> An example test case:
> 
> ```ts
> it('has access to EntityCrudRepository methods', async () => {
>   const note: Note = await repo.create(
>     new Note({title: 'groceries', content: 'eggs,bacon', category: 'keto'}),
>   );
> 
>   const exists = await repo.exists(note.id);
>   expect(exists).to.equal(true);
> });
> ```

In our slack conversation you mentioned `exists` or `findById` since they come from `EntityCrudRepository`. `create` comes from `CrudRepository`, and `findByTitle` from mixin. So I added a test case for : create, findById, and findByTitle. No need for `exists` if we have `findById`. :)

raymondfeng commented on packages/repository/src/__tests__/unit/mixins/add-method-to-repository.mixin.unit.ts:
```diff
@@ -0,0 +1,55 @@
+  // method from CrudRepository
+  it(`non-mixin method 'create' exists`, async () => {
+    expect(repo).to.have.property('create');
+    expect(repo.create).to.be.a.Function();
```
Nitpick: this assertion already covers the one above.

raymondfeng commented on packages/repository/src/__tests__/unit/mixins/add-method-to-repository.mixin.unit.ts:
```diff
@@ -0,0 +1,55 @@
+  // method from EntityCrudRepository
+  it(`non-mixin method 'findById' exists`, async () => {
+    expect(repo).to.have.property('findById');
+    expect(repo.findById).to.be.a.Function();
```
Ditto.

raymondfeng commented on packages/repository/src/__tests__/unit/mixins/add-method-to-repository.mixin.unit.ts:
```diff
@@ -0,0 +1,55 @@
+
+  // method from mixin
+  it(`mixin method 'findByTitle' exists`, async () => {
+    expect(repo).to.have.property('findByTitle');
```
Ditto.

emonddr commented:
hmm. A repo and controller generated from CLI expects EntityCrudRepository methods for the repo. So I need to change the repo mixin again. Will update this tonight.

emonddr commented:
1) Working

If we have

**repo mixin:**

```
export function FindByTitleRepositoryMixin<
  E extends Entity & {title: string},
  R extends Constructor<CrudRepository<E>>
>(superClass: R) {
  class MixedRepository extends superClass implements FindByTitle<E> {
    async findByTitle(title: string): Promise<E[]> {
      const where = {title} as Where<E>;
      const titleFilter = {where};
      return this.find(titleFilter);
    }
  }
  return MixedRepository;
}
```

**note repo:**

```
export class NoteRepository extends FindByTitleRepositoryMixin<
  Note,
  Constructor<
    DefaultCrudRepository<Note, typeof Note.prototype.id, NoteRelations>
  >
>(DefaultCrudRepository) {
  constructor(dataSource: juggler.DataSource) {
    super(Note, dataSource);
  }
}
```

Then repo has access to all CrudRepo and EntityCrudRep methods.

2) Not Working

repo mixin:

```
export function FindByTitleRepositoryMixin<E extends Entity & {title: string}>(
  superClass: Constructor<CrudRepository<E>>,
) {
  return class extends superClass implements FindByTitle<E> {
    async findByTitle(title: string): Promise<E[]> {
      const where = {title} as Where<E>;
      const titleFilter = {where};
      return this.find(titleFilter);
    }
  };
}
```

note repo :  

```
export class NoteRepository extends FindByTitleRepositoryMixin<Note>(
  DefaultCrudRepository,
) {
  constructor(dataSource: juggler.DataSource) {
    super(Note, dataSource);
  }
}
```

Then repo doesn't have access to EntityCrudRepo methods.

emonddr commented:
@bajtos , @raymondfeng  , I got the mixins using Entity instead of Model, and I reverted the repo mixin definition into one where repo can see the CrudRep methods and EntityCrudRepo methods. So if our mocha test for repo mixin can identify the 'create' (CrudRepo), 'findById' (EntityCrudRepo), and findByTitle (mixin), we are good. This also means a CLI generated controller which uses a CLI generated repo (using DefaultCrudRepository) can make use of all the repo methods. Late last week we moved from this repo mixin definition which was working, to a slimmer one we thought worked, but it actually made the entitycrudrepo methods unavailable.

Please take a look at the mixin definitions in the mocha tests directory, and if everyone's happy...I will update the mixin.md file accordingly. thx.

raymondfeng commented:
@emonddr I have suggested changes in https://github.com/strongloop/loopback-next/pull/4778. If you agree, please merge #4778 to your feature branch directly (you don't have to manually update your branch).

bajtos commented on packages/repository/src/__tests__/fixtures/mixins/find-by-title-repo-mixin.ts:
```diff
@@ -0,0 +1,32 @@
+
+export function FindByTitleRepositoryMixin<
+  E extends Entity & {title: string},
+  R extends Constructor<CrudRepository<E>>
```
It seems to me that you are mixing two clean approaches here and arriving at a mixed and a kind of "dirty" solution.

- Approach 1: Require only Model + CRUD methods: use `M extends Model & {title: string}` and `R extends Constructor<CrudRepository<M>>`
- Approach 2: Require `Entity` + CRUD + entity-specific methods, to make the mixin easier to use: use `E extends Entity & {title: string}` and `R extends Constructor<EntityCrudRepository<E>>`.

IIRC, the idea behind the second approach was to simplify definition of `NoteRepository` so that it does not have to specify full repository type as you have now:

```ts
export class NoteRepository extends FindByTitleRepositoryMixin<
   Note,
   Constructor<
     DefaultCrudRepository<Note, typeof Note.prototype.id, NoteRelations>
   >
 >(DefaultCrudRepository)  {
  // ...
}
```

IIRC, we were aiming for the following usage in Approach 2:

```ts
export class NoteRepository extends FindByTitleRepositoryMixin<
   Note,
   typeof Note.prototype.id,
   NoteRelations,
 >(DefaultCrudRepository) {
  // ...
}
```

Based on @raymondfeng proposal in https://github.com/strongloop/loopback-next/pull/4778, I think we should simply revert back to Approach 1 and use `Model` + `CrudRepository` in the generic arguments of the mixin.

emonddr replied:
@bajtos, @raymondfeng 

I cannot seem to make this work:

> R extends Constructor<EntityCrudRepository<E, ID, Relations>>

VSCode complains about ID and Relations.

emonddr replied:
> It seems to me that you are mixing two clean approaches here and arriving at a mixed and a kind of "dirty" solution. 

Not sure why using Entity instead of Model is a dirty approach. Can you clarify?

raymondfeng replied:
With the introduction of `R extends Constructor<CrudRepository<M>>` in your PR, I was able to downgrade `Entity` to `Model` in #4778 and all tests are passing. I think we now have a good solution for the repo mixin to use `Model` directly.

bajtos commented on packages/repository/src/__tests__/unit/mixins/add-method-to-repository.mixin.unit.ts:
```diff
@@ -0,0 +1,52 @@
+
+  // method from CrudRepository
+  it(`non-mixin method 'create' exists`, async () => {
+    expect(repo.create).to.be.a.Function();
```
In fact, calling `repo.create` on the next line is already covering the assertion `expect(repo.create).to.be.a.Function()`. If `create` is not a method (a function), then the JavaScript runtime will throw an error that will fail the test:

```
Uncaught TypeError: repo.create is not a function
```

IMO, the error message reported by the runtime will be actually more helpful than the assertion message, because it can describe both the variable (`repo`) and the member name (`create`). The assertion message will just say something like `expected undefined to be a function`.

I am proposing to remove these `.to.be.a.Function()` assertions from all three tests in this file.

bajtos commented on packages/repository/src/__tests__/unit/mixins/add-method-to-repository.mixin.unit.ts:
```diff
@@ -0,0 +1,52 @@
+
+  // method from EntityCrudRepository
+  it(`non-mixin method 'findById' exists`, async () => {
+    expect(repo.findById).to.be.a.Function();
```
Please remove per my comment above.

bajtos commented on packages/repository/src/__tests__/unit/mixins/add-method-to-repository.mixin.unit.ts:
```diff
@@ -0,0 +1,52 @@
+
+  // method from mixin
+  it(`mixin method 'findByTitle' exists`, async () => {
+    expect(repo.findByTitle).to.be.a.Function();
```
Please remove per my comment above.

bajtos commented on packages/rest/src/__tests__/fixtures/controllers/note.controller.ts:
```diff
@@ -0,0 +1,69 @@
+import {Constructor} from '@loopback/core';
```
Missing license header, please add (see #4778)

bajtos commented on packages/rest/src/__tests__/unit/controller-mixin/controller.mixin.unit.ts:
```diff
@@ -0,0 +1,54 @@
+  });
+
+  it(`mixin method 'findByTitle' exists`, async () => {
+    expect(controller.findByTitle).to.be.a.Function();
```
Similarly here, this check is redundant and produces less helpful failure message than the next line calling `controller.findByTitle`. Please remove.

emonddr commented:
@raymondfeng 

> @emonddr I have suggested changes in #4778. If you agree, please merge #4778 to your feature branch directly (you don't have to manually update your branch).

Actually there are a few spots that `Entity` got changed back to `Model`, so I won't be applying your PR to my feature branch. I've looked at the PR, and have manually copied over the changes. But kept `Entity` instead of changing it to Model. Thank you for the enhancements, though. :)

raymondfeng commented on packages/repository/src/__tests__/fixtures/repositories/note.repository.ts:
```diff
@@ -0,0 +1,28 @@
+// License text available at https://opensource.org/licenses/MIT
+
+import {Constructor} from '@loopback/core';
+import {DataSource} from 'loopback-datasource-juggler';
```
We should avoid direct import to `loopback-datasource-juggler`.  Use `juggler.DataSource` instead.

```
import {juggler, DefaultCrudRepository} from '../../..';
```

raymondfeng commented on packages/rest/package.json:
```diff
@@ -63,7 +63,8 @@
     "@types/qs": "^6.9.1",
-    "multer": "^1.4.2"
+    "multer": "^1.4.2",
+    "loopback-datasource-juggler": "^4.18.1"
```
Please revert this.

raymondfeng commented on packages/rest/src/__tests__/fixtures/repositories/note.repository.ts:
```diff
@@ -0,0 +1,29 @@
+// License text available at https://opensource.org/licenses/MIT
+
+import {DefaultCrudRepository, Where} from '@loopback/repository';
+import {DataSource} from 'loopback-datasource-juggler';
```
Ditto. Use `juggler.DataSource`

emonddr commented on packages/repository/src/__tests__/fixtures/mixins/category-property-mixin.ts:
```diff
@@ -0,0 +1,25 @@
+    category: string;
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    constructor(...args: any[]) {
```
I removed the constructors from all the mixins as they are not needed. tested this. They still work.

jannyHou commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,730 @@ sidebar: lb4_sidebar
+#### Defining A BaseEntity Class Which Extends Entity
+
+Let's define a base model class `BaseEntity` in **src/models/base-entity.ts**.
+It will be used an input to the mixin later.
```
```suggestion
It will be used as input to the mixin later.
```

emonddr replied:
Good catch!

jannyHou commented on docs/site/migration/models/mixins.md:
````diff
@@ -6,8 +6,730 @@ sidebar: lb4_sidebar
+{% include code-caption.html content="src/models/base-entity.ts" %}
+
+```ts
+import {Entity} from '@loopback/repository';
````
(I know it's a rare case and outside of the scope of this PR :p) If we apply the mixin to a `Model` instead of `Entity`, also need to define a BaseModel class to extend the abstract [`Model`](https://github.com/strongloop/loopback-next/blob/master/packages/repository/src/model.ts#L203) right?

emonddr replied:
> (I know it's a rare case and outside of the scope of this PR :p) If we apply the mixin to a `Model` instead of `Entity`, also need to define a BaseModel class to extend the abstract [`Model`](https://github.com/strongloop/loopback-next/blob/master/packages/repository/src/model.ts#L203) right?

Out of scope for this PR. Most people will use Entity and EntityCrudRepository. This guide describes HOW to create a mixin in LB4.  With that knowledge, they can create whatever mixins they want, and targeting whatever model objects they want. :)

jannyHou commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,730 @@ sidebar: lb4_sidebar
+## Creating A Custom Model Method And Remote Model Method Mixin
+
+This section covers the approach that LoopBack 3 can use to add a custom method
+/ remote method to a model via a mixin, and similarly how LoopBack 4 can add a
```
```suggestion
/remote method to a model via a mixin, and similarly how LoopBack 4 can add a
```

jannyHou commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,730 @@ sidebar: lb4_sidebar
+The developer defines a custom model method/remote method mixin in
+**common/mixins/findByTitle.js** which adds a custom method `findByTitle` to any
+model, and adds a corresponding remote method definition with path
+`/findByTitle` as well. An options property `returnArgumentName` makes it
```
```suggestion
`/findByTitle` as well. An option property `returnArgumentName` makes it
```
ignore it if I am wrong :p

emonddr replied:
@jannyHou  : `options` is the name of the parameter. And `returnArgumentName` is a property inside it.

jannyHou commented on docs/site/migration/models/mixins.md:
```diff
@@ -6,8 +6,730 @@ sidebar: lb4_sidebar
+For a full example of a CLI-generated controller for a model `Todo`, see
+[TodoController ](https://github.com/strongloop/loopback-next/blob/master/examples/todo/src/controllers/todo.controller.ts).
+
+#### Adjust NoteController To Use FindByTitleControllerMixin
```
```suggestion
#### Adjusting NoteController To Use FindByTitleControllerMixin
```

emonddr commented:
> looks like mixins are not automatically migrated. We ask developers to create the necessary classes, is it ? should we mention that some where in the notes ?

@deepakrkris , automated commands for migration will possibly come later; depending on our roadmap. At the moment, we are providing manual instructions for migration.
